plugin.js

```diff
@@ -25,9 +26,9 @@ function getName (func) {
   return func.toString().split('\n').slice(0, 2).map(s => s.trim()).join(' -- ')
 }
 
-function Plugin (parent, func, opts, isAfter, timeout) {
+function Plugin (parent, func, optsOrFunc, isAfter, timeout) {
   this.func = func
-  this.opts = opts
+  this.opts = optsOrFunc instanceof Function ? deepClone(optsOrFunc()) : optsOrFunc
```

Comment:
Can you use `typeof` instead of `instanceof`? Same result, better performances :D
```suggestion
  this.opts = typeof optsOrFunc === 'function' ? deepClone(optsOrFunc()) : optsOrFunc
```

Reply:
I thought `instanceof` would be faster as it doesn't do string comparison but after testing I was wrong, there is a whopping 410m ops/sec difference for objects and 440m ops/sec for functions. I will integrate your commit in this push along with the removal of deep cloning.